### PR TITLE
C22- Phoenix- Alejandra G. 

### PR DESCRIPTION
**Pull request by 222alle111**

### CODE REVIEW

File: src/components/ChatEntry.jsx [hunk 1]
```diff
@@ -1,20 +1,34 @@
 import './ChatEntry.css';
+import TimeStamp from './TimeStamp';
+import PropTypes from 'prop-types';
+
+const ChatEntry = (props) => {
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:5]: `ChatEntry` uses a single `props` param, but `ChatLog` uses a destructured object. Personally, I prefer the destructured style, since it makes the expected component attributes a bit more clear. And it's fine to use a mixture of styles in this project, but in general, try to pick one style or the other.

File: src/components/ChatEntry.jsx [hunk 2]
```diff
@@ -1,20 +1,34 @@
...
   return (
     <div className="chat-entry local">
-      <h2 className="entry-name">Replace with name of sender</h2>
+      <h2 className="entry-name">{props.sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{props.body}</p>
+        <p className="entry-time"><TimeStamp time={props.timeStamp} /></p>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:17]: Nice use of the supplied `TimeStamp`. All we need to do is pass in the `timeStamp` string from the message data and it takes care of the rest. All we had to do was confirm the name and type of the prop it was expecting (which we could do through its PropTypes) and we're all set!

File: src/components/ChatEntry.jsx [hunk 3]
```diff
@@ -1,20 +1,34 @@
...
 };
 
 ChatEntry.propTypes = {
   // Fill with correct proptypes
+  id: PropTypes.number.isRequired,
+  sender: PropTypes.string.isRequired,
+  body: PropTypes.string.isRequired,
+  timeStamp: PropTypes.string.isRequired,
+  liked: PropTypes.bool.isRequired,
+  toggleLiked: PropTypes.func.isRequired,
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:31]: The `id`, `sender`, `body`, `timeStamp`, and `liked` props are always passed (they're defined explicitly in the data and also provided in the test) so we can (and should) marked them `isRequired`. 

The remaining props were up to you, and the tests don't know about them. As a result, using `isRequired` causes a warning when running any tests that only pass the known props. If you didn't see those warnings when running the tests, be sure to also try running the terminal `npm test` since the warnings are more visible.

To properly mark any other props `isRequired`, we would also need to update the tests to include at least dummy values (such as an empty callback `() => {}` for the like handler) to make the proptype checking happy.

Alternatively, for any props that we leave not required, we should also have logic in our component to not try to use the value if it's undefined.

File: src/components/ChatLog.jsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+ChatLog.propTypes = {
+  entries: PropTypes.arrayOf(PropTypes.shape({
+    id: PropTypes.number.isRequired,
+    sender: PropTypes.string.isRequired,
+    body: PropTypes.string.isRequired,
+    timeStamp: PropTypes.string.isRequired,
+    liked: PropTypes.bool.isRequired,
+  })).isRequired,
+  toggleLiked: PropTypes.func.isRequired,
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:36]: Similar to the props for `ChatEntry` here, the `entries` prop is included in the tests, but the like toggle is not, resulting in prop warnings (unless we update the tests to reflect our custom props).

Again, if we were to leave this as not required so as to avoid the test warnings, we'd want to be sure that all the script logic in our component worked properly even in the absence of this value.

File: src/components/ChatLog.jsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  const entryComponents = entries.map((entry) => {
+    return (
+      <ChatEntry
+        id={entry.id}
+        sender={entry.sender}
+        body={entry.body}
+        timeStamp={entry.timeStamp}
+        liked={entry.liked}
+        toggleLiked={toggleLiked}
+        key={entry.id}
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:15]: 👍 The `key` attribute is important for React to be able to detect certain kinds of data changes in an efficient manner. We're also using the `id` for our own `id` prop, so it might feel redundant to pass both, but one is for our logic and one is for React internals (we can't safely access the `key` value in any meaningful way).

File: src/components/ChatLog.jsx [hunk 6]
```diff
@@ -0,0 +1,38 @@
+import './ChatLog.css';
+import ChatEntry from './ChatEntry';
+import PropTypes from 'prop-types';
+
+const ChatLog = ({entries, toggleLiked}) => {
+  const entryComponents = entries.map((entry) => {
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:6]: Nice use of `map` to convert from the message data into `ChatEntry` components. We can perform this mapping storing the result into a variable we use in the JSX result as you did here (components are functions, so we can run JS code as usual before we reach the `return`, and even sometimes have multiple `return` statements with different JSX), we could make a helper function that we call as part of the `return`, or this expression itself could be part of the `return` JSX, which I often like since it helps me see the overall structure of the component, though can make debugging a little more tricky. But any of those approaches will work fine.

File: src/components/ChatLog.jsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        id={entry.id}
+        sender={entry.sender}
+        body={entry.body}
+        timeStamp={entry.timeStamp}
+        liked={entry.liked}
+        toggleLiked={toggleLiked}
+        key={entry.id}
+      />);
+  });
+  return (
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:18]: Consider adding a blank line before the `return` to help visually separate the pre-calculation from the JSX result part.

File: src/components/ChatLog.jsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    <>
+      <h2>Chat Log</h2>
+      <ul>
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:21]: Since you're using an `ul` tag to list your chat entries, the child of this should be `li` tags. This can be accomplished either by having `ChatEntry` use `li` as its top-level tag (yours uses `div`) or by wrapping each `ChatEntry` in an `li` in the mapping step. We should aim to generate valid HTML once all the JSX components have been expanded to actual tags.

File: src/App.jsx [hunk 9]
```diff
@@ -1,14 +1,36 @@
...
 const App = () => {
+  const [chatData, setChatData] = useState(chatMessages);
+  const toggleLiked = (id) => {
+    const data = chatData.map(chat => {
+      if (chat.id === id) {
+        return {...chat, liked: !chat.liked};
+      }else {
+        return chat;
+      };
+    });
```
Comment (anselrognlie) [inline src/App.jsx:16]: Nice use of map here to both handle making a new list so that React sees the message data has changed, and make new data for the clicked message with its like status toggled.

File: src/App.jsx [hunk 10]
```diff
@@ -1,14 +1,36 @@
...
+    setChatData(data);
```
Comment (anselrognlie) [inline src/App.jsx:17]: In this case, calculating the next version of the message data using the current state variable and passing that updated version to the state setter shouldn't cause any problems, but we still generally prefer using the callback style of setters. Using that approach, we pass a function to the setter whose parameter will receive the latest state value, and which returns the new value to use for the state.
```js
    setChatData(chatData => chatData.map(chat => {
      if (chat.id === id) {
        return {...chat, liked: !chat.liked};
      } else {
        return chat;
      };
    }));
```

File: src/App.jsx [hunk 11]
```diff
@@ -1,14 +1,36 @@
...
+  };
+  let likeCount = 0;
+  for (const chat of chatData) {
+    if (chat.liked){likeCount += 1};
+  };
```
Comment (anselrognlie) [inline src/App.jsx:22]: Nice job summing the total likes based on the like data of each message. We don't need an additional piece of state to track this, since it can be derived from the existing state we are tracking.

Explicitly totalling the count is perfectly fine, but many JS programmers would use `reduce` to achieve this.
```js
  const likeCount = chatData.reduce((acc, chat) => {
     return chat.liked ? acc + 1 : acc;
  }, 0);
```

The first few times we work with `reduce`, it can be challenging to understand, but it's a tool that gets used commonly enough that it's worth practicing.

File: src/App.jsx [hunk 12]
```diff
@@ -1,14 +1,36 @@
 import './App.css';
+import ChatLog from './components/ChatLog';
+import chatMessages from './data/messages.json';
+import {useState} from 'react';
+
 
 const App = () => {
+  const [chatData, setChatData] = useState(chatMessages);
+  const toggleLiked = (id) => {
```
Comment (anselrognlie) [inline src/App.jsx:9]: 👍 Since our state is defined here, we also need to define our mutating function here so that it can "see" the setter function. All we need to receive is the `id` of the message to toggle, which allows us to locate the message to update, and calculate the next state value. We can pass this all the way down to our `ChatEntry` which handles the click event, passing us the `id` of the message that was clicked.

File: src/components/ChatEntry.jsx [hunk 13]
```diff
@@ -1,20 +1,34 @@
 import './ChatEntry.css';
+import TimeStamp from './TimeStamp';
+import PropTypes from 'prop-types';
+
+const ChatEntry = (props) => {
+  const toggleLikeButton = () => {
+    props.toggleLiked(props.id);
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:7]: 👍 Passing the `id` of this message lets the logic defined up in the `App` find the message to update in its data.

File: src/components/ChatEntry.jsx [hunk 14]
```diff
@@ -1,20 +1,34 @@
...
+  };
+
+  const heartButton= props.liked ? '❤️' : '🤍';
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:10]: 👍 We can figure out which emoji to use for the liked status based on the `liked` prop without creating any additional state.

File: src/components/ChatEntry.jsx [hunk 15]
```diff
@@ -1,20 +1,34 @@
...
     <div className="chat-entry local">
-      <h2 className="entry-name">Replace with name of sender</h2>
+      <h2 className="entry-name">{props.sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{props.body}</p>
+        <p className="entry-time"><TimeStamp time={props.timeStamp} /></p>
+        <button onClick={toggleLikeButton} className="like">{heartButton}</button>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:18]: 👍 We need a wrapper of some kind rather than calling the received callback through props, since our callback function is expecting a message id as its parameter. If we tried to use it directly as the click event handler, React would end up passing it a clink event, since any function registered as an event handler will always be given the event detail information as its argument.

File: src/App.jsx [hunk 16]
```diff
@@ -1,14 +1,36 @@
...
+  };
+  let likeCount = 0;
+  for (const chat of chatData) {
+    if (chat.liked){likeCount += 1};
+  };
   return (
     <div id="App">
       <header>
-        <h1>Application title</h1>
+        <h1>Chat between {chatData[0].sender} and {chatData[1].sender} </h1>
```
Comment (anselrognlie) [inline src/App.jsx:26]: Nice way to read the participant names from the messages. Notice that this assumes there are only two participants in this conversation, and that they are found in the first two messages. What would happen for other conversation situations?
- no one has sent a message yet
- only one participant has sent a message
- a single participant sends multiple messages before getting a response
- there are more than two participants in this conversation

Some of these cases might not really be workable given the limited data we're working with, but it's worth thinking about what this could look like for a more complete application.

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments, and let me know if you have any questions. Nice job!
